File: BasicSamples/BasicTestApp/app/src/main/java/com/kaltura/playerdemo/BaseDemoActivity.java
package com.kaltura.playerdemo;

import android.content.Context;
import android.content.DialogInterface;
import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.ImageView;
import android.widget.ListView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBarDrawerToggle;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.core.view.GravityCompat;
import androidx.drawerlayout.widget.DrawerLayout;

import com.google.android.material.navigation.NavigationView;
import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonSyntaxException;
import com.kaltura.netkit.utils.ErrorElement;
import com.kaltura.playkit.PKDrmParams;
import com.kaltura.playkit.PKLog;
import com.kaltura.playkit.PKMediaEntry;
import com.kaltura.playkit.PKMediaFormat;
import com.kaltura.playkit.PKMediaSource;
import com.kaltura.playkit.PKPluginConfigs;
import com.kaltura.playkit.player.MediaSupport;
import com.kaltura.tvplayer.KalturaPlayer;
import com.kaltura.tvplayer.PlayerInitOptions;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Set;

import static com.kaltura.playkit.PKMediaEntry.MediaEntryType.Unknown;

public abstract class BaseDemoActivity extends AppCompatActivity
        implements NavigationView.OnNavigationItemSelectedListener, KalturaPlayer.OnEntryLoadListener {

    public static final String PLAYER = "player";
    public static final String AUDIO_LANG = "audioLanguage";
    public static final String TEXT_LANG = "textLanguage";
    public static final String PLAYBACK = "playback";
    public static final String OFF = "off";
    public static final String AUTOPLAY = "autoPlay";
    public static final String PRELOAD = "preload";
    //public static final String START_TIME = "startTime";
    public static final String CONFIG = "config";
    public static final String PLUGINS = "plugins";
    public static final String AUTO = "auto";
    public static final String OPTIONS = "options";
    public static final String UICONF_ID ="uiConfId";
    public static final String PARTNER_ID = "partnerId";
    public static final String REFERRER = "referrer";
    public static final String KS = "ks";
    public static final String SERVER_URL = "serverUrl";
    public static final String ALLOW_CROSS_PROTOCOL_ENABLED = "allowCrossProtocolEnabled";
    public static final String STREAM_PRIORITY = "streamPriority";


    private static final PKLog log = PKLog.get("BaseDemoActivity");
    public static final String PLAYER_CONFIG = "playerConfig";
    public static final String UICONF ="uiConf";

    protected final Context context = this;
    protected JsonObject playerConfigUiConfJson;

    protected PlayerInitOptions initOptions;
    String ks;
    DemoItem[] items;
    private ViewGroup contentContainer;
    private NavigationView navigationView;
    private ListView itemListView;

    protected abstract DemoItem[] items();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        initDrm();

        loadConfigFile();

        final DrawerLayout drawer = findViewById(R.id.drawer_layout);
        ActionBarDrawerToggle toggle = new ActionBarDrawerToggle(
                this, drawer, toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close);
        drawer.addDrawerListener(toggle);
        toggle.syncState();

        navigationView = findViewById(R.id.nav_view);
        navigationView.setNavigationItemSelectedListener(this);

        contentContainer = findViewById(R.id.content_container);
        contentContainer.addView(getItemListView());
        navigationView.setCheckedItem(R.id.nav_gallery);
    }

    protected void parseInitOptions(JsonObject json) {
        final Integer partnerId = safeInteger(json, PARTNER_ID);
        if (partnerId == null) {
            throw new IllegalArgumentException("partnerId must not be null");
        }
        if (json.has(PLAYER_CONFIG)) {
            JsonObject playerConfigJasonObject = safeObject(json, PLAYER_CONFIG);
            final PlayerInitOptions options = new PlayerInitOptions(partnerId);
            if (initOptions == null) {
                options.setAutoPlay(safeBoolean(playerConfigJasonObject, AUTOPLAY))
                        .setPreload(safeBoolean(playerConfigJasonObject, PRELOAD))
                        .setKs(safeString(playerConfigJasonObject, KS))
                        .setPluginConfigs(parsePluginConfigs(json.get(PLUGINS)))
                        .setAllowCrossProtocolEnabled(safeBoolean(playerConfigJasonObject, ALLOW_CROSS_PROTOCOL_ENABLED))
                        .setReferrer(safeString(playerConfigJasonObject, REFERRER));

            }
            initOptions = options;
        }
    }

    private PKPluginConfigs parsePluginConfigs(JsonElement json) {
        PKPluginConfigs configs = new PKPluginConfigs();
        if (json != null && json.isJsonObject()) {
            final JsonObject obj = json.getAsJsonObject();
            for (Map.Entry<String, JsonElement> entry : obj.entrySet()) {
                final String pluginName = entry.getKey();
                final JsonElement value = entry.getValue();
                configs.setPluginConfig(pluginName, value);
            }
        }
        return configs;
    }

    protected JsonObject safeObject(JsonObject json, String key) {
        final JsonElement jsonElement = json.get(key);
        if (jsonElement != null && jsonElement.isJsonObject()) {
            return jsonElement.getAsJsonObject();
        }
        return null;
    }

    protected String safeString(JsonObject json, String key) {
        final JsonElement jsonElement = json.get(key);
        if (jsonElement != null && jsonElement.isJsonPrimitive()) {
            return jsonElement.getAsString();
        }
        return null;
    }

    protected Boolean safeBoolean(JsonObject json, String key) {
        final JsonElement jsonElement = json.get(key);
        if (jsonElement != null && jsonElement.isJsonPrimitive()) {
            return jsonElement.getAsBoolean();
        }
        return null;
    }

    protected Integer safeInteger(JsonObject json, String key) {
        final JsonElement jsonElement = json.get(key);
        if (jsonElement != null && jsonElement.isJsonPrimitive()) {
            return jsonElement.getAsInt();
        }
        return null;
    }

    protected abstract void loadConfigFile();

    void initDrm() {
        MediaSupport.initializeDrm(this, new MediaSupport.DrmInitCallback() {
            @Override
            public void onDrmInitComplete(Set<PKDrmParams.Scheme> supportedDrmSchemes, boolean provisionPerformed, Exception provisionError) {
                if (provisionPerformed) {
                    if (provisionError != null) {
                        log.e("DRM Provisioning failed", provisionError);
                    } else {
                        log.d("DRM Provisioning succeeded");
                    }
                }
                log.d("DRM initialized; supported: " + supportedDrmSchemes);

                // Now it's safe to look at `supportedDrmSchemes`
            }
        });
    }

    protected void parseCommonOptions(JsonObject json) {
        parseInitOptions(safeObject(json, "initOptions"));

        if (initOptions != null) {
            ks = initOptions.ks;
        }
        final JsonArray jsonItems = json.get("items").getAsJsonArray();
        List<DemoItem> itemList = new ArrayList<>(jsonItems.size());
        for (JsonElement item : jsonItems) {
            final JsonObject object = item.getAsJsonObject();
            itemList.add(parseItem(object));
        }

        items = itemList.toArray(new DemoItem[itemList.size()]);
    }

    protected void parseBasicCommonOptions(JsonObject json) {
        parseInitOptions(safeObject(json, "initOptions"));

        if (initOptions != null) {
            ks = initOptions.ks;
        }
        final JsonArray jsonItems = json.get("items").getAsJsonArray();
        List<DemoItem> itemList = new ArrayList<>(jsonItems.size());
        for (JsonElement item : jsonItems) {
            final JsonObject object = item.getAsJsonObject();
            PKMediaEntry pkMediaEntry = MockMediaParser.parseMedia(object);
            itemList.add(new DemoItem(pkMediaEntry.getName(), pkMediaEntry.getId(), pkMediaEntry));
        }

        items = itemList.toArray(new DemoItem[itemList.size()]);
    }

    static class MockMediaParser {

        static PKMediaEntry parseMedia(JsonObject mediaObject) throws JsonSyntaxException {

            PKMediaEntry mediaEntry = new Gson().fromJson(mediaObject, PKMediaEntry.class);
            if (mediaEntry.getMediaType() == null) {
                mediaEntry.setMediaType(Unknown);
            }
            List<PKMediaSource> mediaSources = mediaEntry.getSources();
            for (PKMediaSource mediaSource : mediaSources) {
                PKMediaFormat format = PKMediaFormat.valueOfUrl(mediaSource.getUrl());
                if (format == null) {
                    String mimeType = getMimeTypeFromJson(mediaObject);
                    if (mimeType != null) {
                        if (mimeType.equals(PKMediaFormat.dash.mimeType)) {
                            format = PKMediaFormat.dash;
                        } else if (mimeType.equals(PKMediaFormat.hls.mimeType)) {
                            format = PKMediaFormat.hls;
                        } else if (mimeType.equals(PKMediaFormat.wvm.mimeType)) {
                            format = PKMediaFormat.wvm;
                        } else if (mimeType.equals(PKMediaFormat.mp4.mimeType)) {
                            format = PKMediaFormat.mp4;
                        } else if (mimeType.equals(PKMediaFormat.mp3.mimeType)) {
                            format = PKMediaFormat.mp3;
                        }
                    }
                }
                mediaSource.setMediaFormat(format);
            }
            return mediaEntry;
        }
    }

    @Nullable
    private static String getMimeTypeFromJson(JsonObject mediaObject) {
        String mimeType = null;
        try {
            JSONObject jsonObj = new JSONObject(mediaObject.toString());
            JSONArray sources = jsonObj.getJSONArray("sources");
            if (sources != null && sources.length() > 0) {
                JSONObject sourcesJson = sources.getJSONObject(0);
                mimeType = sourcesJson.getString("mimeType");
                return mimeType;
            }
        } catch (JSONException e) {
            //e.printStackTrace();
            log.d("Sources does not contain mime type in it - hope url extension is valid...");
        }
        return mimeType;
    }

    @NonNull
    protected abstract DemoItem parseItem(JsonObject object);

    protected int partnerId() {
        if (initOptions == null || initOptions.tvPlayerParams == null) {
            return 0;
        }
        return initOptions.tvPlayerParams.partnerId;
    }

    protected abstract void loadItem(DemoItem item);

    @Override
    public void onEntryLoadComplete(PKMediaEntry entry, ErrorElement error) {
        if (error != null) {
            Log.d("onEntryLoadComplete", " error: " + error);
        }
    }

    @Override
    public void onBackPressed() {
        DrawerLayout drawer = findViewById(R.id.drawer_layout);
        if (drawer.isDrawerOpen(GravityCompat.START)) {
            drawer.closeDrawer(GravityCompat.START);
        } else {
            super.onBackPressed();
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.main, menu);

        TextView partnerIdTextView = findViewById(R.id.partnerIdTextView);
        partnerIdTextView.setText("Partner: " + partnerId());

        TextView demoNameTextView = findViewById(R.id.demoNameTextView);
        demoNameTextView.setText(demoName());

        ImageView icon = findViewById(R.id.imageView);
        icon.setColorFilter(android.R.color.holo_green_dark);

        return true;
    }

    protected abstract String demoName();

    @Override
    public boolean onNavigationItemSelected(@NonNull MenuItem item) {
        // Handle navigation view item clicks here.
        contentContainer.removeAllViews();

        switch (item.getItemId()) {
            case R.id.nav_gallery:
                contentContainer.addView(getItemListView());
                break;
            case R.id.nav_downloads:
            case R.id.nav_plugins:
                // TODO
                break;
        }

        DrawerLayout drawer = findViewById(R.id.drawer_layout);
        drawer.closeDrawer(GravityCompat.START);
        return true;
    }

    private ListView getItemListView() {
        if (itemListView != null) {
            return itemListView;
        }

        ArrayAdapter<DemoItem> itemArrayAdapter = new ArrayAdapter<>(this, android.R.layout.simple_list_item_1);
        itemArrayAdapter.addAll(items());


        itemListView = new ListView(this);

        itemListView.setAdapter(itemArrayAdapter);

        itemListView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(final AdapterView<?> parent, View view, final int position, long id) {
                new AlertDialog.Builder(context)
                        .setTitle(getString(R.string.select_action))
                        .setItems(new String[]{getString(R.string.play_stream)}, new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialog, int which) {
                                if (which == 0) {
                                    loadItem(((DemoItem) parent.getItemAtPosition(position)));
                                }
                            }
                        }).show();
            }
        });

        return itemListView;
    }
}

File: BasicSamples/BasicTestApp/app/src/main/java/com/kaltura/playerdemo/BaseDemoActivity.kt
package com.kaltura.playerdemo

import android.content.Context
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.ViewGroup
import android.widget.*
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.core.view.GravityCompat
import androidx.drawerlayout.widget.DrawerLayout
import com.google.android.material.navigation.NavigationView
import com.google.gson.Gson
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import com.google.gson.JsonSyntaxException
import com.kaltura.netkit.utils.ErrorElement
import com.kaltura.playkit.PKLog
import com.kaltura.playkit.PKMediaEntry
import com.kaltura.playkit.PKMediaEntry.MediaEntryType.Unknown
import com.kaltura.playkit.PKMediaFormat
import com.kaltura.playkit.PKPluginConfigs
import com.kaltura.playkit.player.MediaSupport
import com.kaltura.tvplayer.KalturaPlayer
import com.kaltura.tvplayer.PlayerInitOptions
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.content_main.*
import org.json.JSONException
import org.json.JSONObject
import java.util.*

abstract class BaseDemoActivity : AppCompatActivity(), NavigationView.OnNavigationItemSelectedListener, KalturaPlayer.OnEntryLoadListener {

    protected val context: Context = this
    protected var playerConfigUiConfJson: JsonObject? = null

    protected var initOptions: PlayerInitOptions? = null
    var ks: String? = ""
    var items: Array<DemoItem>? = null
    private var contentContainer: ViewGroup? = null
    private var itemListView: ListView? = null

    protected abstract fun items(): Array<DemoItem>

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)

        initDrm()

        loadConfigFile()

        val drawer = findViewById<DrawerLayout>(R.id.drawer_layout)
        val toggle = ActionBarDrawerToggle(
                this, drawer, toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close)
        drawer.addDrawerListener(toggle)
        toggle.syncState()

        nav_view.setNavigationItemSelectedListener(this)

        content_container.addView(getItemListView())
        nav_view.setCheckedItem(R.id.nav_gallery)
    }

    protected fun parseInitOptions(json: JsonObject?) {
        val partnerId = safeInteger(json, PARTNER_ID)
                ?: throw IllegalArgumentException("partnerId must not be null")
        if (json?.has(PLAYER_CONFIG)!!) {
            val playerConfigJasonObject = safeObject(json, PLAYER_CONFIG)
            val options = PlayerInitOptions(partnerId)
            if (initOptions == null) {
                options.setAutoPlay(safeBoolean(playerConfigJasonObject!!, AUTOPLAY))
                        .setPreload(safeBoolean(playerConfigJasonObject, PRELOAD))
                        .setKs(safeString(playerConfigJasonObject, KS))
                        .setPluginConfigs(parsePluginConfigs(json.get(PLUGINS)))
                        .setAllowCrossProtocolEnabled(safeBoolean(playerConfigJasonObject, ALLOW_CROSS_PROTOCOL_ENABLED))
                        .setReferrer(safeString(playerConfigJasonObject, REFERRER))

            }
            initOptions = options
        }
    }

    private fun parsePluginConfigs(json: JsonElement?): PKPluginConfigs {
        val configs = PKPluginConfigs()
        if (json != null && json.isJsonObject) {
            val obj = json.asJsonObject
            for ((pluginName, value) in obj.entrySet()) {
                configs.setPluginConfig(pluginName, value)
            }
        }
        return configs
    }

    protected fun safeObject(json: JsonObject?, key: String): JsonObject? {
        val jsonElement = json?.get(key)
        return if (jsonElement != null && jsonElement.isJsonObject) {
            jsonElement.asJsonObject
        } else null
    }

    protected fun safeString(json: JsonObject, key: String): String? {
        val jsonElement = json.get(key)
        return if (jsonElement != null && jsonElement.isJsonPrimitive) {
            jsonElement.asString
        } else null
    }

    protected fun safeBoolean(json: JsonObject, key: String): Boolean? {
        val jsonElement = json.get(key)
        return if (jsonElement != null && jsonElement.isJsonPrimitive) {
            jsonElement.asBoolean
        } else null
    }

    protected fun safeInteger(json: JsonObject?, key: String): Int? {
        val jsonElement = json?.get(key)
        return if (jsonElement != null && jsonElement.isJsonPrimitive) {
            jsonElement.asInt
        } else null
    }

    protected abstract fun loadConfigFile()

    private fun initDrm() {
        MediaSupport.initializeDrm(this) { pkDeviceSupportInfo, provisionError ->
            if (pkDeviceSupportInfo.isProvisionPerformed) {
                if (provisionError != null) {
                    log.e("DRM Provisioning failed", provisionError)
                } else {
                    log.d("DRM Provisioning succeeded")
                }
            }
            log.d("DRM initialized; supported: ${pkDeviceSupportInfo.supportedDrmSchemes} isHardwareDrmSupported: ${pkDeviceSupportInfo.isHardwareDrmSupported}")

            // Now it's safe to look at `supportedDrmSchemes`
        }
    }

    protected fun parseCommonOptions(json: JsonObject) {
        parseInitOptions(safeObject(json, "initOptions"))

        ks = initOptions?.ks

        val jsonItems = json.get("items").asJsonArray
        val itemList = ArrayList<DemoItem>(jsonItems.size())
        for (item in jsonItems) {
            val `object` = item.asJsonObject
            itemList.add(parseItem(`object`))
        }

        items = itemList.toTypedArray()
    }

    protected fun parseBasicCommonOptions(json: JsonObject) {
        parseInitOptions(safeObject(json, "initOptions"))

        ks = initOptions?.ks

        val jsonItems = json.get("items").asJsonArray
        val itemList = ArrayList<DemoItem>(jsonItems.size())
        for (item in jsonItems) {
            val `object` = item.asJsonObject
            val pkMediaEntry = MockMediaParser.parseMedia(`object`)
            itemList.add(DemoItem(pkMediaEntry.name, pkMediaEntry.id, pkMediaEntry))
        }

        items = itemList.toTypedArray()
    }

    internal object MockMediaParser {

        @Throws(JsonSyntaxException::class)
        fun parseMedia(mediaObject: JsonObject): PKMediaEntry {

            val mediaEntry = Gson().fromJson(mediaObject, PKMediaEntry::class.java)
            if (mediaEntry.mediaType == null) {
                mediaEntry.mediaType = Unknown
            }
            val mediaSources = mediaEntry.sources
            for (mediaSource in mediaSources) {
                var format = PKMediaFormat.valueOfUrl(mediaSource.url)
                if (format == null) {
                    val mimeType = getMimeTypeFromJson(mediaObject)
                    if (mimeType != null) {
                        if (mimeType == PKMediaFormat.dash.mimeType) {
                            format = PKMediaFormat.dash
                        } else if (mimeType == PKMediaFormat.hls.mimeType) {
                            format = PKMediaFormat.hls
                        } else if (mimeType == PKMediaFormat.wvm.mimeType) {
                            format = PKMediaFormat.wvm
                        } else if (mimeType == PKMediaFormat.mp4.mimeType) {
                            format = PKMediaFormat.mp4
                        } else if (mimeType == PKMediaFormat.mp3.mimeType) {
                            format = PKMediaFormat.mp3
                        }
                    }
                }
                mediaSource.mediaFormat = format
            }
            return mediaEntry
        }
    }

    protected abstract fun parseItem(`object`: JsonObject): DemoItem

    protected fun partnerId(): Int {
        initOptions?.let {
            it.tvPlayerParams?.let { tvPlayerParam ->
                return tvPlayerParam.partnerId
            }
        }
        return 0
    }

    protected abstract fun loadItem(item: DemoItem)

    override fun onEntryLoadComplete(entry: PKMediaEntry, error: ErrorElement?) {
        if (error != null) {
            Log.d("onEntryLoadComplete", " error: $error")
        }
    }

    override fun onBackPressed() {
        val drawer = findViewById<DrawerLayout>(R.id.drawer_layout)
        if (drawer.isDrawerOpen(GravityCompat.START)) {
            drawer.closeDrawer(GravityCompat.START)
        } else {
            super.onBackPressed()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(R.menu.main, menu)

        val partnerIdTextView = findViewById<TextView>(R.id.partnerIdTextView)
        partnerIdTextView.text = "Partner: " + partnerId()

        val demoNameTextView = findViewById<TextView>(R.id.demoNameTextView)
        demoNameTextView.text = demoName()

        val icon = findViewById<ImageView>(R.id.imageView)
        icon.setColorFilter(android.R.color.holo_green_dark)

        return true
    }

    protected abstract fun demoName(): String

    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        // Handle navigation view item clicks here.
        contentContainer?.removeAllViews()

        when (item.itemId) {
            R.id.nav_gallery -> contentContainer?.addView(getItemListView())
            R.id.nav_downloads, R.id.nav_plugins -> {
            }
        }// TODO

        val drawer = findViewById<DrawerLayout>(R.id.drawer_layout)
        drawer.closeDrawer(GravityCompat.START)
        return true
    }

    private fun getItemListView(): ListView {
        itemListView?.let {
            return it
        }

        val itemArrayAdapter = ArrayAdapter<DemoItem>(this, android.R.layout.simple_list_item_1)
        itemArrayAdapter.addAll(*items())

        itemListView = ListView(this)

        itemListView?.adapter = itemArrayAdapter

        itemListView?.onItemClickListener = AdapterView.OnItemClickListener { parent, view, position, id ->
            AlertDialog.Builder(context)
                    .setTitle(getString(R.string.select_action))
                    .setItems(arrayOf(getString(R.string.play_stream))) { dialog, which ->
                        if (which == 0) {
                            loadItem(parent.getItemAtPosition(position) as DemoItem)
                        }
                    }.show()
        }

        return itemListView!!
    }

    companion object {

        val PLAYER = "player"
        val AUDIO_LANG = "audioLanguage"
        val TEXT_LANG = "textLanguage"
        val PLAYBACK = "playback"
        val OFF = "off"
        val AUTOPLAY = "autoPlay"
        val PRELOAD = "preload"
        //public static final String START_TIME = "startTime";
        val CONFIG = "config"
        val PLUGINS = "plugins"
        val AUTO = "auto"
        val OPTIONS = "options"
        val UICONF_ID = "uiConfId"
        val PARTNER_ID = "partnerId"
        val REFERRER = "referrer"
        val KS = "ks"
        val SERVER_URL = "serverUrl"
        val ALLOW_CROSS_PROTOCOL_ENABLED = "allowCrossProtocolEnabled"
        val STREAM_PRIORITY = "streamPriority"


        private val log = PKLog.get("BaseDemoActivity")
        val PLAYER_CONFIG = "playerConfig"
        val UICONF = "uiConf"

        private fun getMimeTypeFromJson(mediaObject: JsonObject): String? {
            var mimeType: String? = null
            try {
                val jsonObj = JSONObject(mediaObject.toString())
                val sources = jsonObj.getJSONArray("sources")
                if (sources != null && sources.length() > 0) {
                    val sourcesJson = sources.getJSONObject(0)
                    mimeType = sourcesJson.getString("mimeType")
                    return mimeType
                }
            } catch (e: JSONException) {
                //e.printStackTrace();
                log.d("Sources does not contain mime type in it - hope url extension is valid...")
            }

            return mimeType
        }
    }
}
